x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SetSingleNodeAllocateStep.java

```diff
@@ -69,49 +79,127 @@ public void performAction(IndexMetadata indexMetadata, ClusterState clusterState
         final Map<ShardId, List<ShardRouting>> routingsByShardId = clusterState.getRoutingTable()
             .allShards(indexName)
             .stream()
             .collect(Collectors.groupingBy(ShardRouting::shardId));
 
+
         if (routingsByShardId.isEmpty() == false) {
+            List<String> allNodeIds = new ArrayList<>();
             for (RoutingNode node : routingNodes) {
-                boolean canAllocateOneCopyOfEachShard = routingsByShardId.values().stream() // For each shard
-                    .allMatch(shardRoutings -> shardRoutings.stream() // Can we allocate at least one shard copy to this node?
-                        .map(shardRouting -> allocationDeciders.canAllocate(shardRouting, node, allocation).type())
-                        .anyMatch(Decision.Type.YES::equals));
-                if (canAllocateOneCopyOfEachShard) {
-                    validNodeIds.add(node.node().getId());
-                }
+                allNodeIds.add(node.node().getId());
             }
-            // Shuffle the list of nodes so the one we pick is random
-            Randomness.shuffle(validNodeIds);
-            Optional<String> nodeId = validNodeIds.stream().findAny();
+            NodesStatsRequest nodesStatsRequest = new NodesStatsRequest(allNodeIds.toArray(new String[0])).clear().
+                addMetric(NodesStatsRequest.Metric.FS.metricName()).indices(new CommonStatsFlags(CommonStatsFlags.Flag.Store));
+            getClient().admin().cluster().nodesStats(nodesStatsRequest, ActionListener.wrap((nodesStatsResponse) -> {
+                List<RoutingNode> validRoutingNodes = new ArrayList<>();
+                final Map<String, Long> nodeShardsStorageBytes = new HashMap<>();
+
+                Map<String, NodeStats> nodeStatsMap = nodesStatsResponse.getNodesMap();
+                for (String nodeId : allNodeIds) {
+                    if (nodeStatsMap.get(nodeId) != null) {
+                        List<IndexShardStats> indexShardStatsList = nodeStatsMap.get(nodeId).getIndices().
+                            getShardStats(indexMetadata.getIndex());
+                        long shardsOnCurrentNodeStorageBytes = indexShardStatsList.stream().mapToLong(indexShardStats ->
+                            Arrays.stream(indexShardStats.getShards()).mapToLong(shardStats ->
+                                shardStats.getStats().getStore().getSizeInBytes()).sum()).sum();
+                        if (shardsOnCurrentNodeStorageBytes != 0) {
+                            nodeShardsStorageBytes.put(nodeId, shardsOnCurrentNodeStorageBytes);
+                        }
+                    }
+                }
+                long indexPrimaryShardsStorageBytes = nodeShardsStorageBytes.values().stream().mapToLong(Long::longValue).sum();
+                if (indexMetadata.getNumberOfReplicas() != 0) {
+                    indexPrimaryShardsStorageBytes /= indexMetadata.getNumberOfReplicas();
+                }
```

Comment:
This can be removed if we use the primary stats above

Reply:
The calculation above is used to indicate "how much of the source index's storage" is available per node. As shrink works from both primary and replica I believe it's correct to include replicas in the math.